### PR TITLE
Fix outdated data references in derived glyphs

### LINKED ISSUE #14629: `HoverTool` highlights outdated data when dynamically changing `MultiLine` data column in source
### Software versions

Python version        :  3.13.5 (main, Jul 11 2025, 22:43:46) [Clang 20.1.4 ]
IPython version       :  (not installed)
Tornado version       :  6.5.2
NumPy version         :  2.3.2
Bokeh version         :  3.8.0
BokehJS static path   :  /home/ediun/git/joy/.venv/lib/python3.13/site-packages/bokeh/server/static
node.js version       :  (not installed)
npm version           :  (not installed)
jupyter_bokeh version :  (not installed)
Operating system      :  Linux-6.6.87.1-microsoft-standard-WSL2-x86_64-with-glibc2.35

### Browser name and version

_No response_

### Jupyter notebook / Jupyter Lab version

_No response_

### Expected behavior

When adding a `HoverTool` on a `MultiLine` glyph, and then dynamically changing the data column to display through a JS callback (e.g., through a button), the hover tool should highlight the new data.

### Observed behavior

Updating the data causes the highlighting to appear on previous (no longer displayed) data (seemingly on the same row as the new data). Moreover, the line supposed to be highlighted in the new data disappears while hovering it.

### Example code

```Python
import bokeh.layouts as bl
import bokeh.models as bm
import bokeh.plotting as bp

source = bm.ColumnDataSource(
    {
        "ys": [[1, 2, 1]],
        "xs1": [[1, 2, 3]],
        "xs2": [[4, 5, 6]],
    }
)

figure = bp.figure(x_range=bm.Range1d(0, 7))
lines = figure.multi_line(
    "xs1", "ys", line_width=3, hover_line_color="red", source=source
)

hover = bm.HoverTool(tooltips=None)
figure.add_tools(hover)

button = bm.Button(label="Change columns")
button.js_on_click(
    bm.CustomJS(
        args={"lines": lines},
        code="""
            lines.glyph.xs.field = "xs2";
            lines.glyph.change.emit();
        """,
    )
)

layout = bl.layout([figure, button])
bp.show(layout)
```

### Stack traceback or browser console output

_No response_

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by muendlein**
- [x] issues: fixes #14629
- [x] tests added / passed

### CODE REVIEW

File: bokehjs/src/lib/models/glyphs/glyph.ts [hunk 1]
```diff
@@ -302,6 +303,11 @@ export abstract class GlyphView extends DOMComponentView {
   set_base<T extends this>(base: T): void {
     if (base != this && base instanceof this.constructor) {
       this._base = base
+      for (const prop of this.model) {
+        if (prop instanceof p.BaseCoordinateSpec) {
```
Comment (canavandl) [inline bokehjs/src/lib/models/glyphs/glyph.ts:307]: Does this issue only affect "x" and "y" properties? I feel like updating a visual property like "fill_color" would have the same issue.

File: bokehjs/src/lib/models/glyphs/glyph.ts [hunk 2]
```diff
@@ -294,6 +294,7 @@ export abstract class GlyphView extends DOMComponentView {
   }
 
   protected _base: this | null = null
+  protected _should_inherit_from_BaseCoordinateSpec: boolean = true
```
Comment (canavandl) [inline bokehjs/src/lib/models/glyphs/glyph.ts:297]: nit: this logic is hard for me to grasp. Would it be better to either not cache ala

```

// Just make the decision at the point of use:
protected _should_inherit<T>(prop: p.Property<T>, base: this): boolean {
  // DataSpec properties (coordinates, colors, etc.) should always inherit
  // to track field reference changes
  if (prop instanceof p.DataSpec) {
    return true
  }
  
  // Other properties only inherit if values match
  return this._can_inherit_from(prop, base)
}
```

or rename to something like

```
// Or describes the behavior:
_force_coordinate_inheritance: boolean = false

// Or most explicit:
_coordinates_need_forced_inheritance: boolean = false
```

File: bokehjs/src/lib/models/glyphs/glyph.ts (hunk 1 above)
Comment (muendlein) [reply bokehjs/src/lib/models/glyphs/glyph.ts:307]: For visual properties this is actually not an issue but rather the desired behavior. Changing the color of the base glyph should not sync back the corresponding glyph.

File: bokehjs/src/lib/models/glyphs/glyph.ts (hunk 2 above)
Comment (muendlein) [reply bokehjs/src/lib/models/glyphs/glyph.ts:297]: Agreed, names require an update.

The local caching is actually on purpose, as the enforced data field inheritance requires some finer control. Specifically there is a test with decoupled data fields when the glyph renderer is initialized. In this PR the coupling is defined when the `base` is set. One thing currently missing is a reset of the forced inheritance to false when the data field of the derived glyph gets directly updated. However I'm not sure if there is even a use case for this edge case.
Apart from that a partial coupling of just x or y is also not included.

https://github.com/bokeh/bokeh/blob/8741a11ed77ccaab6d683b7af038cb18b99c07c3/bokehjs/test/integration/glyphs/glyphs.ts#L1074-L1095

**Comment by bryevdv** [comment]
@muendlein I have not been in this code in some time so I asked Codex for a review. Let me know what you think 

> #### [P1] Runtime overrides of derived coordinates are ignored
> 
> Once `_should_inherit_from_BaseCoordinateSpec` becomes `true`, it is never cleared. If a hover or selection glyph initially shares `xs` with the base glyph and later explicitly changes its own `xs`, `set_data()` still inherits the base array and discards that override.
> 
> This regresses the existing ability to update derived glyph coordinates dynamically. The inheritance state needs to be invalidated when the derived property changes.
> 
> [Relevant code](https://github.com/bokeh/bokeh/blob/6b3caa4d973b1f4a99671e83dbe9ffcfcc805837/bokehjs/src/lib/models/glyphs/glyph.ts#L458-L460)
> 
> #### [P1] Coordinate inheritance is tracked glyph-wide instead of per property
> 
> `set_base()` ANDs every coordinate comparison into one boolean. If a derived glyph intentionally overrides `x` but initially shares `y`, the differing `x` disables forced inheritance for `y` too.
> 
> A subsequent base `y` field change therefore leaves the derived glyph using stale `y` data—the same class of bug this PR is intended to fix. Inheritance should be tracked per attribute, with a regression test covering a partial `x`/`y` override.
> 
> [Relevant code](https://github.com/bokeh/bokeh/blob/6b3caa4d973b1f4a99671e83dbe9ffcfcc805837/bokehjs/src/lib/models/glyphs/glyph.ts#L303-L308)
